src/VisualStudio/Core/Def/Implementation/Workspace/SourceGeneratedFileManager.cs

```diff
@@ -355,30 +356,29 @@ public async Task UpdateBufferContentsAsync(CancellationToken cancellationToken)
                 }
                 else
                 {
+                    var generatedDocuments = await project.GetSourceGeneratedDocumentsAsync(cancellationToken).ConfigureAwait(false);
+                    var generatedDocument = generatedDocuments.SingleOrDefault(d =>
+                                d.SourceGenerator.GetType().FullName.Equals(_generatorTypeName, StringComparison.OrdinalIgnoreCase) &&
+                                d.SourceGenerator.GetType().Assembly.GetName().Name.Equals(_generatorAssemblyName));
 
-                    var generatorDriverRunResults = await project.GetGeneratorDriverRunResultAsync(cancellationToken).ConfigureAwait(false);
-                    var generatorRunResult = generatorDriverRunResults?.Results.SingleOrNull(r =>
-                                r.Generator.GetType().FullName.Equals(_generatorTypeName, StringComparison.OrdinalIgnoreCase) &&
-                                r.Generator.GetType().Assembly.GetName().Name.Equals(_generatorAssemblyName));
-
-                    if (generatorRunResult == null)
+                    if (generatedDocument != null)
                     {
-                        windowFrameMessageToShow = string.Format(ServicesVSResources.The_generator_0_that_generated_this_file_has_been_removed_from_the_project, _generatorTypeName);
-                        windowFrameImageMonikerToShow = KnownMonikers.StatusError;
+                        windowFrameMessageToShow = string.Format(ServicesVSResources.This_file_is_autogenerated_by_0_and_cannot_be_edited, _generatorTypeName);
+                        windowFrameImageMonikerToShow = default;
+                        generatedSource = await generatedDocument.GetTextAsync(cancellationToken).ConfigureAwait(false);
                     }
                     else
                     {
-                        generatedSource = generatorRunResult.Value.GeneratedSources.SingleOrNull(r => r.HintName == _generatedSourceHintName);
-
-                        if (generatedSource == null)
+                        // The file isn't there anymore; do we still have the generator at all?
+                        if (project.AnalyzerReferences.Any(a => a.GetGenerators(project.Language).Any(g => g.GetType().Assembly.GetName().Name.Equals(_generatorAssemblyName))))
```

Comment:
perhaps extract out a helper for the assembly name bit, as that logic shows up in a few places and should likel be uniform.